### ISSUE TITLE
Rails swallows exceptions when loading models/controllers in certain cases

### ISSUE THREAD

**Issue by janko** [rails/rails#38431]
### Steps to reproduce

```
$ rails new my_app
$ cd my_app
$ rails g scaffold posts
$ rails db:migrate
```

```rb
# app/models/post.rb
class Post < ApplicationRecord
  foo # cause an error
end
```

```rb
# app/controllers/posts_controller.rb
class PostsController < ApplicationController
  Post # reference the model on the class-level
  # ... other actions ...
end
```

### Expected behavior

When I visit a page in the rails controller, I expected Rails to show me a `NoMethodError` exception that `Post.foo` method doesn't exist, coming from `app/models/post.rb`.

### Actual behavior

When Rails starts loading `app/controllers/posts_controller.rb`, and encounters the reference to `Post` model (which it will fail to load), it will silently stop evaluating the code and continue rendering the page.

In this case, it seems that `app/views/posts/index.html.erb` will still render even if `PostsContoller#index` was never explicitly defined (the evaluation of our controller never reached `def index`). So I get the following exception:

```
ActionView::Template::Error (undefined method `each' for nil:NilClass):
    10:   </thead>
    11:
    12:   <tbody>
    13:     <% @posts.each do |post| %>
    14:       <tr>
    15:         <td><%= link_to 'Show', post %></td>
    16:         <td><%= link_to 'Edit', edit_post_path(post) %></td>

app/views/posts/index.html.erb:13
```

because the implicit `PostsController#index` action doesn't define the `@posts` instance variable.

### System configuration
**Rails version**: 6.0.2.1

**Ruby version**: 2.6.5

**Comment by kaspth** [comment]
cc @fxn 

Thanks for the report 🙏

**Comment by eugeneius** [comment]
This was fixed in https://github.com/rails/rails/pull/37834, which will be released in 6.0.3.

**Comment by fxn** [comment]
> When Rails starts loading app/controllers/posts_controller.rb, and encounters the reference to Post model (which it will fail to load), it will silently stop evaluating the code and continue rendering the page.

Let me clarify that the one that stops evaluating the code and continues execution is Ruby itself, not Rails.

`class C` defines the `C` class, and if the class definition body raises, Ruby does not undo anything, because from the point of view of Ruby the interpreter is just executing code top to bottom. If the caller catches the exception, `C` is half-defined. That is Ruby semantics.
